extensions/ql-vscode/src/codeql-cli/cli.ts

```diff
@@ -419,6 +421,32 @@ export class CodeQLCliServer implements Disposable {
           // Listen to stderr
           process.stderr.addListener("data", (newData: Buffer) => {
             stderrBuffers.push(newData);
+
+            if (!silent) {
+              currentLineStderrBuffers.push(newData);
+
+              // Print the stderr to the logger as it comes in. We need to ensure that
+              // we don't split messages on the same line, so we buffer the stderr and
+              // split it on EOLs.
+              let currentLineBuffer = Buffer.concat(currentLineStderrBuffers);
+              const eolBuffer = Buffer.from(EOL);
+              if (currentLineBuffer.includes(eolBuffer)) {
+                while (currentLineBuffer.includes(eolBuffer)) {
+                  const line = currentLineBuffer.subarray(
+                    0,
+                    currentLineBuffer.indexOf(eolBuffer),
+                  );
+
+                  void this.logger.log(line.toString("utf-8"));
+
+                  currentLineBuffer = currentLineBuffer.subarray(
+                    currentLineBuffer.indexOf(eolBuffer) + eolBuffer.length,
+                  );
+                }
+
+                currentLineStderrBuffers = [currentLineBuffer];
+              }
```

Comment:
We check whether the buffer contains `eolBuffer` or calculate the index quite a lot here. Once first of all, then once as the loop condition, and then twice within the loop. I recognise it's not a very expensive operation of the buffer is in-memory and likely to be short, but I wonder if it's worth avoiding anyway.

I normally find it confusing the set a variable in the loop condition, but this could be a case where it works quite well. Then we calculate it only once per loop.

If you also go with my earlier suggestion of converting the array of buffers into a single buffer, the code could look like:
```suggestion
              let eolIndex;
              while (
                (eolIndex = currentLineStderrBuffer.indexOf(eolBuffer)) !== -1
              ) {
                const line = currentLineStderrBuffer.subarray(0, eolIndex);
                void this.logger.log(line.toString("utf-8"));

                currentLineStderrBuffer = currentLineStderrBuffer.subarray(
                  eolIndex + eolBuffer.length,
                );
              }
```

Reply:
Thanks, I think that makes sense. I did find that because we're doing `subarray` here, we also need to copy the buffer to prevent memory leaks (because a subarray keeps a reference to the original buffer), but this was also present in my original code.